packages/gatsby/src/utils/merge-gatsby-config.ts

```diff
@@ -46,15 +32,43 @@ const howToMerge = {
    * This makes sure that if a single value is defined, that one it used.
    * We prefer the "right" value, because the user's config will be "on the right"
    */
-  byDefault: (a, b) => b || a,
-  siteMetadata: (objA, objB) => _.merge({}, objA, objB),
+  byDefault: (a: ConfigKey, b: ConfigKey): ConfigKey => b || a,
+  siteMetadata: (objA: Metadata, objB: Metadata): Metadata =>
+    _.merge({}, objA, objB),
   // plugins are concatenated and uniq'd, so we don't get two of the same plugin value
-  plugins: (a = [], b = []) =>
+  plugins: (a: Plugins = [], b: Plugins = []): Plugins =>
     _.uniqWith(a.concat(b), (a, b) =>
       _.isEqual(
         _.pick(normalizePluginEntry(a), [`resolve`, `options`]),
         _.pick(normalizePluginEntry(b), [`resolve`, `options`])
       )
     ),
-  mapping: (objA, objB) => _.merge({}, objA, objB),
+  mapping: (objA: Mapping, objB: Mapping): Mapping => _.merge({}, objA, objB),
+}
+
+/**
+ * Defines how a theme object is merged with the user's config
+ */
+export const mergeGatsbyConfig = (a, b): IGatsbyConfig => {
```

Comment:
`a` and `b` should also be typed as well

Now that I think about it - we perhaps need 2 interfaces for config:
- one for input that allow all various ways users can define plugins (as string, as object with just `resolve`, and as object with both `resolve` and `options`) - essentially all the ways users can define in their `gatsby-config` - this would be type for `a` and `b` arguments and likely the return as well as we don't really normalize here - just merge stuff
- second one for normalized one (ensuring existance of fields, shape of entries in `plugins` array, etc) - likely not used here because we still don't normalize config just yet (at least not the `plugins` part of it)

I still think we should not import type from root `index.d.ts` (as we will want later on go the other way around and in index.d.ts we will likely export appropriate internal types) - we however can copy it here for now (the gatsby config input type doesn't end up in redux so it shouldn't be in same place as the redux type). Alternatively - instead of copying we can move it here and re-export it in `index.d.ts`

Reply:
Thanks for reviewing! I agree with it.

While I resolve this conversation, I found that NormalizedPlugin used for just comparison. It doesn't change user config normalized.
So I can't type NormalizedConfig now.